### PR TITLE
Support string concatenation for java 11

### PR DESCRIPTION
**Pull request by yuvaraj-anbarasan**
Summary of my contribution during Google Summer of Code:
```
Compiling the module classes using --patch-module in gradle
Handling LambdaMetaFactory
Added modules to classPath
Fixed setBootstrapMethod
Added missing methods in Unsafe and JPF_jdk_internal_misc_Unsafe classes
Fixed setting wrong reference value
Fix for String Concatenation
Fixed issue of loading classes from jar-file
Fixed error in URLClassLoaderTest.testFindResource
Test cases for string concatenation
Fixed string concatenation with empty strings
Fixed concatenation of String and Byte
Fixed concatenation of String and char, int, float
```

Improved test success rate from 20% to 90%.

### CODE REVIEW

File: src/main/gov/nasa/jpf/jvm/JVMClassInfo.java [hunk 1]
```diff
@@ -116,38 +116,63 @@ public void setBootstrapMethodCount (ClassFile cf, Object tag, int count) {
...
-      String samDescriptor = cf.methodTypeDescriptorAt(cpArgs[2]);
-            
-      if(lambdaBody!=null) {
-        bootstrapMethods[idx] = new BootstrapMethodInfo(lambdaRefKind, JVMClassInfo.this, lambdaBody, samDescriptor);
+      if(methodBody!=null) {
+        bootstrapMethods[idx] = new BootstrapMethodInfo(refKind, JVMClassInfo.this, methodBody, descriptor
+                , bmArg, bmType);
       }
     }
-
```
Comment (cyrille-artho) [inline src/main/gov/nasa/jpf/jvm/JVMClassInfo.java]: Avoid changes in whitespace

File: src/main/gov/nasa/jpf/jvm/JarClassFileContainer.java [hunk 2]
```diff
@@ -115,11 +115,16 @@ static String getPath(File file, String pathPrefix){
...
   public ClassFileMatch getMatch(String clsName) throws ClassParseException {
     String classEntryURL = getClassEntryURL(clsName);
+
+    if(clsName.startsWith("java."))
+     classEntryURL = "modules" + "/" +classEntryURL;
     
     if (pathPrefix != null){
       classEntryURL = pathPrefix + classEntryURL;
     }
-
```
Comment (cyrille-artho) [inline src/main/gov/nasa/jpf/jvm/JarClassFileContainer.java]: Avoid changes in whitespace

File: src/main/gov/nasa/jpf/jvm/JarClassFileContainer.java [hunk 3]
```diff
@@ -131,7 +136,7 @@ public ClassFileMatch getMatch(String clsName) throws ClassParseException {
         }
 
         is = jar.getInputStream(e);
-
```
Comment (cyrille-artho) [inline src/main/gov/nasa/jpf/jvm/JarClassFileContainer.java]: Avoid changes in whitespace

File: src/main/gov/nasa/jpf/vm/DynamicElementInfo.java [hunk 4]
```diff
@@ -130,11 +130,13 @@ public String asString() {
    */
   @Override
   public byte[] getStringBytes(){
+
```
Comment (cyrille-artho) [inline src/main/gov/nasa/jpf/vm/DynamicElementInfo.java]: Avoid changes in whitespace

File: src/main/gov/nasa/jpf/vm/DynamicElementInfo.java [hunk 5]
```diff
@@ -130,11 +130,13 @@ public String asString() {
    */
   @Override
   public byte[] getStringBytes(){
+  
     if (!ClassInfo.isStringClassInfo(ci)) {
       throw new JPFException("object is not of type java.lang.String");
     }
 
     int valueFieldRef = getDeclaredReferenceField("value", "java.lang.String");
+
```
Comment (cyrille-artho) [inline src/main/gov/nasa/jpf/vm/DynamicElementInfo.java]: Avoid changes in whitespace

File: src/main/gov/nasa/jpf/vm/ElementInfo.java [hunk 6]
```diff
@@ -1152,6 +1152,10 @@ public double getDoubleField (FieldInfo fi){
     }
   }
   public int getReferenceField (FieldInfo fi) {
+    // System.err.println("/**********************************/");
```
Comment (cyrille-artho) [inline src/main/gov/nasa/jpf/vm/ElementInfo.java]: remove

File: src/main/gov/nasa/jpf/vm/FunctionObjectFactory.java [hunk 7]
```diff
@@ -17,28 +17,70 @@
  */
 package gov.nasa.jpf.vm;
 
-/**
+/*
```
Comment (cyrille-artho) [inline src/main/gov/nasa/jpf/vm/FunctionObjectFactory.java]: add "*" back

File: src/tests/gov/nasa/jpf/test/java/net/URLClassLoaderTest.java [hunk 8]
```diff
@@ -6,13 +6,13 @@
  * The Java Pathfinder core (jpf-core) platform is licensed under the
  * Apache License, Version 2.0 (the "License"); you may not use this file except
  * in compliance with the License. You may obtain a copy of the License at
- *
```
Comment (cyrille-artho) [inline src/tests/gov/nasa/jpf/test/java/net/URLClassLoaderTest.java]: Avoid changes in whitespace

File: src/tests/gov/nasa/jpf/test/java/net/URLClassLoaderTest.java [hunk 9]
```diff
@@ -231,7 +237,6 @@ public void testFindResource() throws MalformedURLException {
       url = cl.findResource(resInterface1);
       expectedUrl = jarUrl + resInterface1;
       assertEquals(url.toString(), expectedUrl);
-
```
Comment (cyrille-artho) [inline src/tests/gov/nasa/jpf/test/java/net/URLClassLoaderTest.java]: Avoid changes in whitespace

**Comment by cyrille-artho** [comment]
Looks good, although a few of the extra whitespace changes are still there (but it's relatively tedious to track them down and fix them all).
Merging is now possible.
One last thing:
Travis currently fails, because the .travis.yml in this branch does not work with the current setup of Travis:
```
apt-get install failed

The command "sudo -E apt-get -yq --no-install-suggests --no-install-recommends $(travis_apt_get_options) install oracle-java11-installer oracle-java11-set-default" failed and exited with 100 during .
```
Can you please look into this?

**Comment by yuvaraj-anbarasan** [comment]
I'll check it out

**Pull request merged**